Fix deleting lockup account and Staking page

marcinbodnar opened the pull request:
Error described in #2245 fixed by changing how `lockupIdExists` in `getProfileStakingDetails` action is set. Additionally handled `staking.getLockup` in `account` reducer to set `hasLockup` value that will help to reason what should be done while setting `lockupIdExists`.`hasLockup` also used in `StakingContainer` component.

Additionally, I found another bug. The problem was that if the user wanted to enter the Staking page after the lockup account was deleted, Wallet crashed. That was because we are storing previously selected `accountId` for Staking page in the local storage - and if the user has previously selected lockup account, we were trying to load staking details for this lockup account - that was failing. 

Fixed it by adding functionality that will detect if the lockup account stored in local storage exists, and if not it will switch to the main account automatically.

esaminu commented:
@marcinbodnar The second issue you found sounds very similar to https://github.com/near/near-wallet/issues/2210. Is it the same issue? I'm writing that test case now.

MaximusHaximus commented on packages/frontend/src/redux/reducers/account/index.js:
```diff
@@ -199,7 +199,16 @@ const account = handleActions({
+    [staking.getLockup]: (state, { error, payload, ready }) => 
+        (!ready || error)
+            ? {
+                ...state 
```
Shouldn't we just return `state` here rather than shallow copying it?

marcinbodnar replied:
Right. Done.

MaximusHaximus reviewed (changes requested):
Added a few things in-line, this code is tricky, <3 :)

marcinbodnar commented on packages/frontend/src/redux/actions/account.js:
```diff
@@ -43,7 +43,7 @@ export const getProfileStakingDetails = (accountId) => async (dispatch, getState
     const lockupIdExists = accountId
         ? !!getState().allAccounts[accountId].balance.lockedAmount
-        : !!getState().account.balance.lockedAmount;
+        : !!getState().account.hasLockup;
```
Yes, all selectors for `account.js` are ready here: #2223 , I would prefer not to mix these changes. I will change it in the #2223 or #2247 depends on which one will be merged first :)

MaximusHaximus replied:
You created this selector _in this PR_ -- let's not continue referencing state directly in new code, ever.
Any new code you write that uses a selector now is one less piece of code that will need to be audited, changed, and tested later.

marcinbodnar replied:
Sure. Done.

marcinbodnar commented on packages/frontend/src/redux/actions/staking.js:
```diff
@@ -432,5 +433,17 @@ export const updateStaking = (currentAccountId, recentlyStakedValidators) => asy
     }
 
-    dispatch(staking.updateCurrent(currentAccountId || accountId));
+    let currentAccount = getState().staking.accounts.find((account) => account.accountId === currentAccountId);
```
Yes, changed it. But the same as with `account.js`, I'm planning to change all selectors for staking actions in one PR. This way I can prepare it in a more systematic way, which allows me to create a bit better implementation.

MaximusHaximus replied:
As above, we are no longer writing direct state references off of the root state; all new state accesses should use selectors.  The less code we add now with direct state references, the less code we need to worry about auditing changing, and testing in the very near future.

marcinbodnar replied:
Sure. It's done.

marcinbodnar commented on packages/frontend/src/redux/actions/account.js:
```diff
@@ -43,7 +43,7 @@ export const getProfileStakingDetails = (accountId) => async (dispatch, getState
 
     const lockupIdExists = accountId
         ? !!getState().allAccounts[accountId].balance.lockedAmount
```
So this check is for external accounts (if you use `/profile/:accountId`) only. I changed the arg name so it will be a bit more readable.

MaximusHaximus replied:
Understood that the data is different, but why do we _not_ need the same fix for 'external accounts'? I would assume that we're going to be showing incorrect lockup account details for 'external' accounts unless we make a similar change to `allAccounts[accountId]` to track `hasLockup` under similar circumstances as the primary account.

marcinbodnar replied:
It's working properly, we don't need `hasLockup` because if we are showing an external account profile, this check:
`!!getState().allAccounts[externalAccountId].balance.lockedAmount`
is working properly and if there is no lockup the `hasLockup` is `undefined`.

Added `hasLockup` to the `allAccounts` reducer, it will be more consistent and more readable.

Generally, the problem is with handling `balance` object in `account` reducer (that was also causing the main problem in this issue) -  it's a bit complicated implementation, probably because there were a million changes to it, while the project was expanding by new functionalities. We should check the balance more deeply, probably create a separate reducer for it, and refactor it step by step - but it's a bit bigger work, if you prefer I can handle it now in separate PR, but the other things would have to wait a bit?

marcinbodnar commented:
> @marcinbodnar The second issue you found sounds very similar to #2210. Is it the same issue? I'm writing that test case now.

@esaminu Yes, added proper comment into #2210

marcinbodnar commented:
@MaximusHaximus @esaminu Additionally fix the bug, where lockup account balance details and transfer from lockup banner were displayed after removing lockup.

MaximusHaximus reviewed (changes requested):
Responded to comments in-line :)

MaximusHaximus reviewed (changes requested):
Added comment replies to your prior comments -- sorry, they got added as raw comments rather than review comments.

esaminu reviewed (approved):
Tests passing and banner + lockup balances now dismissing as expected 👍

marcinbodnar commented:
> Tests passing and banner + lockup balances now dismissing as expected 👍

@esaminu Could you please test it again? I made bigger changes.

esaminu commented:
@marcinbodnar Banner and lockup balances no longer dismissing on latest changes. I also added a small fix for cases where all the remaining balance is in "Reserved for Storage" since currently the contract panics [here](https://github.com/near/core-contracts/blob/master/lockup/src/owner.rs#L469) in this case.

esaminu commented:
I have another comment re [this logic](https://github.com/near/near-wallet/pull/2247/files#diff-46f345c520c144e0d8300842ff78d012a61770029e9d262aaf62632d95960486R66). I think we should call the transfers voting contract directly to check if transfers are enabled and if they are then spend the gas to update the lockup by calling `check_transfers_vote` and then proceed with `transfer` and `deletion` logic. If they aren't, we should return at that point. This logic assumes that all transfers vote contracts have transfers enabled. This should usually be the case with our conditional rendering of the banner but I don't think we should rely on that.

esaminu commented:
I have another thought re: [this logic](https://github.com/near/near-wallet/blob/97e8b1c60ebce416004a4448b18bc5bd8ec5207b/packages/frontend/src/utils/account-with-lockup.js#L66). I think we should call the transfers voting contract directly to check if transfers are enabled and if they are then spend the gas to update the lockup by calling check_transfers_vote and then proceed with transfer and deletion logic. If they aren't, we should return at that point. This logic assumes that all transfers vote contracts have transfers enabled. This should usually be the case with our conditional rendering of the banner but I don't think we should rely on that.

marcinbodnar commented:
@esaminu 
> Banner and lockup balances no longer dismissing on latest changes though owner's balance is updating.
Fixed it, please check.

> I also added a small fix for cases where all the remaining balance is in "Reserved for Storage" since currently the contract panics here in this case.
Great, thanks.

marcinbodnar commented:
@esaminu 
> I have another thought re: [this logic](https://github.com/near/near-wallet/blob/97e8b1c60ebce416004a4448b18bc5bd8ec5207b/packages/frontend/src/utils/account-with-lockup.js#L66). I think we should call the transfers voting contract directly to check if transfers are enabled and if they are then spend the gas to update the lockup by calling `check_transfers_vote` and then proceed with `transfer` and `deletion` logic. If they aren't, we should return / throw at that point. This logic assumes that all transfers vote contracts have transfers enabled. This should usually be the case with our conditional rendering of the banner but I don't think we should rely on that.

It makes sense, but I'm not sure if there aren't any corner cases regarding the current solution - I wasn't involved in creating the current implementation of `transferAllFromLockup` (cc @MaximusHaximus ?)

esaminu commented:
@marcinbodnar 👍  banner is dismissing now but I still see the loading overlay for lockup accounts while navigating around until I reload
<img width="350" alt="Screen Shot 2021-11-19 at 8 54 57 PM" src="https://user-images.githubusercontent.com/5849204/142667707-770d8470-348d-47fb-8264-4e9bd2bbf967.png">
 
@MaximusHaximus @marcinbodnar With current logic, if transfers are disabled we will spend gas to call [`check_transfers_vote`](https://github.com/near/core-contracts/blob/master/lockup/src/owner.rs#L428) and then [`transferAllFromLockup`](https://github.com/near/near-wallet/blob/97e8b1c60ebce416004a4448b18bc5bd8ec5207b/packages/frontend/src/utils/account-with-lockup.js#L63) function execution will end there since other conditions to transfer or delete are false because [`get_locked_amount`](https://github.com/near/core-contracts/blob/master/lockup/src/getters.rs#L65) will return whole amount as locked.

marcinbodnar commented:
@esaminu 
> 👍 banner is dismissing now but I still see the loading overlay for lockup accounts while navigating around until I reload

Thanks, it's fixed.

There ware a conflict regarding your commit from this PR, and your other commit from #2248 - I resolved it by leaving changes from #2248 . Please check.

esaminu commented:
All looks good and passing now. I re added the fix from earlier 👍 

> There was a conflict regarding your commit from this PR, and your other commit from #2248 - I resolved it by leaving changes from #2248 . Please check.

Changes from #2248 seem to still be there.

MaximusHaximus commented:
@marcinbodnar as soon as you've got the conflicts here resolved, please feel free to merge this PR :)

MaximusHaximus commented on packages/frontend/src/utils/account-with-lockup.js:
```diff
@@ -92,16 +92,18 @@ export async function transferAllFromLockup(missingAmount) {
     }
 
     console.info('Attempting to transfer from lockup account ID:', lockupAccountId);
```
this log should be inside of the if check so we don't inaccurately log that we're trying to transfer when we actually aren't due to there being a 0 liquid balance.

MaximusHaximus commented on packages/frontend/src/redux/actions/staking.js:
```diff
@@ -407,9 +412,12 @@ const handleGetAccounts = () => async (dispatch, getState) => {
     try {
-        await dispatch(staking.getLockup(accountId || getState().account.accountId));
+        await dispatch(staking.getLockup({
+            accountId: externalAccountId || getState().account.accountId,
+            isOwner: !externalAccountId || externalAccountId === getState().account.accountId
```
:hocho: Please replace these direct state accesses with selector usages :)

MaximusHaximus reviewed (changes requested):
Just two small things added in-line, as soon as those are handled, :shipit: I think we're ready to get this sucker shipped out! :)

marcinbodnar commented:
@esaminu There were a lot of conflicts while rebasing, I tested it, and looks fine, could you please test it as well, I just want to be sure if it's working properly.

esaminu commented:
@marcinbodnar tests for lockups are ready on https://github.com/near/near-wallet/tree/lockup-e2e-tests branch. Could you try to run them directly? If you have trouble please let me know :)
